### ISSUE TITLE
2.0.0 DI cannot resolve IDBContextOptions

### ISSUE THREAD

**Issue by caleblloyd** [aspnet/EntityFramework#8818]
I am in the process of upgrading [Pomelo.EntityFrameworkCore.MySql](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/tree/dev) to 2.0.0

I'm pretty sure this is a usage issue, but I cannot get past it.  I'm having issues with Dependency Injection resolving `IDBContextOptions` when running `dotnet ef migrations add initial`:

```
System.ArgumentNullException: Value cannot be null.
Parameter name: contextOptions
   at Microsoft.EntityFrameworkCore.Utilities.Check.NotNull[T](T value, String parameterName)
   at Microsoft.EntityFrameworkCore.Storage.RelationalConnectionDependencies..ctor(IDbContextOptions contextOptions, IDiagnosticsLogger`1 transactionLogger, IDiagnosticsLogger`1 connectionLogger, INamedConnectionStringResolver connectionStringResolver)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.MySqlServiceCollectionExtensions.<>c.<AddEntityFrameworkMySql>b__0_2(IServiceProvider p) in /home/caleb/Projects/caleb/Pomelo.EntityFrameworkCore.Mysql/src/EFCore.MySql/Extensions/MySqlServiceCollectionExtensions.cs:line 45
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitTransient(TransientCallSite transientCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.EnsureServices(IServiceProvider services)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.AddMigration(String name, String outputDir, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.AddMigrationImpl(String name, String outputDir, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.<>c__DisplayClass3_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
```

I don't see in the code where IDBContextOptions gets added as scoped/singleton.  Would someone mind pointing me in the right direction?

We are tracking EFCore `2.0.0-preview3-*` from the MyGet feed.  If you want to check out our 2.0.0 code it's on the [dev branch](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/tree/dev)

**Comment by Kagamine** [comment]
@divega

**Comment by ajcvickers** [comment]
@caleblloyd What does your `AddEntityFrameworkMySql` method look like? (I couldn't find one in your dev branch.)

**Comment by caleblloyd** [comment]
@ajcvickers https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/dev/src/EFCore.MySql/Extensions/MySqlServiceCollectionExtensions.cs

**Comment by ajcvickers** [comment]
@caleblloyd It is registered as scoped in EntityFrameworkServicesBuilder, so your call to `TryAddCoreServices` should register it.

Are you seeing this in normal runtime situations, or only in `dotnet ef` tooling scenarios?

**Comment by caleblloyd** [comment]
I haven't been able to try normal runtime situations yet (can't generate a migration) but it happens in all `dotnet ef` situations

**Comment by ajcvickers** [comment]
@caleblloyd What happens if you try a normal runtime scenario? Does it get to the point of querying the database and then fail, or does it fail with this first?

**Comment by caleblloyd** [comment]
@ajcvickers turns out `AddEntityFrameworkMySql` is not being called at all in the app.  Did something change with `2.0.0` that would cause the `Startup.cs` to no longer run when invoked by `dotnet ef`?

- [Program.cs](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/dev/test/EFCore.MySql.FunctionalTests/Program.cs)
- [Startup.cs](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/dev/test/EFCore.MySql.FunctionalTests/Startup.cs)

**Comment by ajcvickers** [comment]
@caleblloyd Yep. See #8328 and it's linked issues.

**Comment by caleblloyd** [comment]
Fantastic, I'll check it out

**Comment by caleblloyd** [comment]
So we actually do call `UseMySql` in `DbContext.OnConfiguring`.  I verified that [UseMySql is called here in OnConfiguring](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/101e058e7c3837d828abbefebf80a98dbc97d1ba/test/EFCore.MySql.FunctionalTests/AppDb.cs#L70)

When it is called in `OnConfiguring`, we get this DI error cannot resolve IDBContextOptions.

When I add this to the DbContext however, the error goes away:

```
		AppDb IDesignTimeDbContextFactory<AppDb>.CreateDbContext(string[] args)
		{
			var optionsBuilder = new DbContextOptionsBuilder<AppDb>()
				.UseMySql(AppConfig.Config["Data:ConnectionString"], options => options.MaxBatchSize(AppConfig.EfBatchSize));
			return new AppDb(optionsBuilder.Options);
		}
```

I feel like the first scenario with `UseMySql` called in `DbContext.OnConfiguring` should work, no?

**Comment by ajcvickers** [comment]
@caleblloyd Is this still only an issue with `dotnet ef...` commands, or in normal runtime scenarios as well? if the latter, could you post the stack trace?

**Comment by caleblloyd** [comment]
@ajcvickers I guess the problem is we call `UseMySql` in this case but `AddEntityFrameworkCoreMySql` is not called.

I read https://github.com/aspnet/EntityFramework/issues/8328 and also read linked issues.  I can't find any good examples.  Do you guys have a good sample for SQL Server somewhere?  Specifically, I'd like to see examples for a .NET Core MVC App and a basic Console app.

This is going to be a huge change for everyone who has an app out there and tries to move to 2.0.0.  I think an example of how to correctly perform this migration from the EF team would help a lot, so we can send people to the example when they open issues to us.

**Comment by ajcvickers** [comment]
@caleblloyd So are you saying that it only happens for dotnet ef commands? If so, @bricelam would be the best one to help.

With regard to the switch from ConfigureServices to BuildWebHost, that isn't something driven by the EF team, but rather by ASP.NET Core. @glennc or @DamianEdwards may have some guidance about migrating applications from the old pattern to the new.

**Comment by caleblloyd** [comment]
It also happens in basic console apps with no DI or `SerciceCollection`.

Did `UseMySql` or `UseSqlServer` used to implicitly configure services in 1.x? Basic console apps with no DI used to work without calling `AddEntityFrameworkXXX`

**Comment by ajcvickers** [comment]
Ultimately the `ApplyServices` method will be called on your OptionsExtension. That will cause the services to be registered in the internal service provider. Except in special cases, application code should not be calling the AddEntityFrameworkXxx methods directly--this was also the case for 1.x releases. The ASP.NET Core item templates stopped using a called to AddEntityFrameworkXxx prior to the 1.0 RTM release.

**Comment by ajcvickers** [comment]
@caleblloyd https://blog.oneunicorn.com/2016/11/03/implementing-a-provider-use-method-for-ef-core-1-1/ is probably the best reference. Some of this has changed for 2.0, but the concepts are still pretty much the same.

**Comment by caleblloyd** [comment]
When the DBContext is configured in `Startup.cs` in the `ConfigureServices` function for a .NET MVC app, using a call to `services.AddDbContext<AppDb>`, everything works fine.  [Here is the configuration that works](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/da21634afa4fcd1dfc403223fec538730b44c712/test/EFCore.MySql.FunctionalTests/Startup.cs#L57-L60)

When the DBContext is configured for a basic CLI app and the `dotnet ef` commands using the DBContext's `OnConfiguring` method, I get the `null IDbContextOptions contextOptions` exception.  [Here is the configuration that throws the exception](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/da21634afa4fcd1dfc403223fec538730b44c712/test/EFCore.MySql.FunctionalTests/AppDb.cs#L80-L83)

In both cases, [ApplyServices was called](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/da21634afa4fcd1dfc403223fec538730b44c712/src/EFCore.MySql/Infrastructure/Internal/MySqlOptionsExtension.cs#L27-L33)  (I added debug logging to verify)

Here is the exception stack trace:

```
dotnet run testMigrate
constructor parameterless called
UseMySql called
ApplyServices called
Value cannot be null.
Parameter name: contextOptions
   at Microsoft.EntityFrameworkCore.Utilities.Check.NotNull[T](T value, String parameterName)
   at Microsoft.EntityFrameworkCore.Storage.RelationalConnectionDependencies..ctor(IDbContextOptions contextOptions, IDiagnosticsLogger`1 transactionLogger, IDiagnosticsLogger`1 connectionLogger, INamedConnectionStringResolver connectionStringResolver)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.MySqlServiceCollectionExtensions.<>c.<AddEntityFrameworkMySql>b__0_2(IServiceProvider p) in /home/caleb/Projects/caleb/Pomelo.EntityFrameworkCore.Mysql/src/EFCore.MySql/Extensions/MySqlServiceCollectionExtensions.cs:line 45
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.EntityFrameworkCore.Infrastructure.AccessorExtensions.GetService[TService](IInfrastructure`1 accessor)
   at Microsoft.EntityFrameworkCore.Infrastructure.DatabaseFacade.get_DatabaseCreator()
   at Microsoft.EntityFrameworkCore.Infrastructure.DatabaseFacade.EnsureDeleted()
   at Pomelo.EntityFrameworkCore.MySql.PerfTests.Commands.TestMigrateCommand.Run() in /home/caleb/Projects/caleb/Pomelo.EntityFrameworkCore.Mysql/test/EFCore.MySql.FunctionalTests/Commands/TestMigrateCommand.cs:line 12
   at Pomelo.EntityFrameworkCore.MySql.PerfTests.Commands.CommandRunner.Run(String[] args) in /home/caleb/Projects/caleb/Pomelo.EntityFrameworkCore.Mysql/test/EFCore.MySql.FunctionalTests/Commands/CommandRunner.cs:line 29
```

I'm not ruling out the fact that this may be an implementation issue on our end, but I'm not sure where to look at this point.  This issue is blocking further work on the 2.0.0 migration of our provider and your help is very much appreciated!

**Comment by ajcvickers** [comment]
@caleblloyd So I cloned and built the dev branch, but I get this exception when attempting to use with EF Core 2.0 dev branch.
![image](https://user-images.githubusercontent.com/1430078/27200906-778db80a-51d0-11e7-8244-0170e39ee066.png)

Is there some other branch I should be pulling?

**Comment by caleblloyd** [comment]
@ajcvickers no, the `dev` branch is correct.  I'm not sure why it's referring to the assembly `Pomelo.EntityFrameworkCore.MySql, Version=1.1.1.0` in your screenshot.  I [upped the Version Prefix](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/dev/src/EFCore.MySql/EFCore.MySql.csproj#L4) to 2.0.0 on `dev`.

**Comment by ajcvickers** [comment]
@caleblloyd Oops. Looks like I was using master by mistake. Let me try again.

**Comment by ajcvickers** [comment]
@caleblloyd So I found one problem. This line:
```C#
.TryAdd<IRelationalTypeMapper>(p => p.GetService<IMySqlScopedTypeMapper>())
```
means that a singleton service is indirectly depending on a scoped service. If the provider needs to change the lifetime of a service, then it will need to do the registration in the TryAddProviderSpecificServices section:
```C#
.TryAddScoped<IRelationalTypeMapper>(p => p.GetService<IMySqlScopedTypeMapper>())
```
This should only be done with extreme caution since it may have cascading effects on other services. For example, anything that depends on IRelationalTypeMapper will now also need to be made scoped, and anything that depends on those services will likewise have to change, and so on. Such changes can have a big impact on the way the stack behaves.

I do remember there was a reason for this, but can you remind me why the type mapper needs to be scoped? Generally speaking it is designed to be a singleton service.

**Comment by caleblloyd** [comment]
It's really meant to be the other way around- the scoped service depends on the Singleton service. 

The scoped type mapper takes into account the MySql server version and connection string options and changes some type mappings. Since an application may have multiple different databases we've made it scoped. [Here is the class](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/master/src/Pomelo.EntityFrameworkCore.MySql/Storage/Internal/MySqlScopedTypeMapper.cs)

Is this the proper use or are different Singleton services instantiated for each DBContext? If they are, we can make the two type mappers back into one Singleton class

**Comment by ajcvickers** [comment]
@caleblloyd Yes, there is. If an option will result in a singleton service behaving differently, then it should be implemented as an ISingletonOptions. These options will be initialized once per service provider and are registered as singletons so that singleton services can depend on them.

Coupled with this is a mechanism for EF to build a new service provider if an option has changed--to do this, implement IDbContextOptionsExtension.GetServiceProviderHashCode to return a different hash when your singleton options change.

There are a couple of examples of this in the EF code. (Although I just noticed that the one for RowNumberPagingEnabled is not doing the correct thing for the service provider hash--I'll get this fixed.)

**Comment by ajcvickers** [comment]
Also, it doesn't seem like the github repo is making use of the functional tests that we publish for providers to use--unless you are doing it in some way that I missed. I would recommend that you look at how the SQL Compact provider (https://github.com/ErikEJ/EntityFramework.SqlServerCompact) and the Prostgres provider (https://github.com/npgsql/Npgsql.EntityFrameworkCore.PostgreSQL) are doing this. I'm sure either @ErikEJ or @roji would be happy to give some guidance on this.

**Comment by caleblloyd** [comment]
> There are a couple of examples of this in the EF code. (Although I just noticed that the one for RowNumberPagingEnabled is not doing the correct thing for the service provider hash--I'll get this fixed.)

Could you add a reference to this issue when you create the PR?  I'd like to see how it's done in SqlServer and copy that behavior.

> Also, it doesn't seem like the github repo is making use of the functional tests that we publish for providers to use

It's on our backlog, thanks for the references.  Would be great if we could add that.  Right now we use our own set of functional tests but using the official one would give us better coverage.

**Comment by ajcvickers** [comment]
@caleblloyd See https://github.com/aspnet/EntityFramework/pull/8878

**Comment by caleblloyd** [comment]
@ajcvickers I made the TypeMapper a singleton int https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/commit/5dd4b6c34696fb7c9da374940b6b4d7a9ca43f6b but I still get the same error with `IDbContextOptions contextOptions` being null for `Microsoft.EntityFrameworkCore.Storage.RelationalConnectionDependencies`

**Comment by caleblloyd** [comment]
I did some advanced debugging and it looks like in the services setup, the class `RelationalConnectionDependencies` is being instantiated twice.

The first time it is instantiated with `IDbContextOptions`:

![image](https://user-images.githubusercontent.com/2414837/27311796-7f1feba6-5531-11e7-85c1-46ce6527c293.png)

The second time it is instantiated `IDbContextOptions` is `null`:

![image](https://user-images.githubusercontent.com/2414837/27311897-1a21b15c-5532-11e7-9155-2ae17924fabc.png)

It appears to me that the crash is coming when the [Database Creator is being loaded](https://github.com/aspnet/EntityFramework/blob/0ebcf4991b335cd9f6bff687e4939434b5dfb6e8/src/EFCore.Relational/Infrastructure/EntityFrameworkRelationalServicesBuilder.cs#L136)

Here are the relevant bits of the stack trace:

```
System.ArgumentNullException: Value cannot be null.
Parameter name: contextOptions
   at Microsoft.EntityFrameworkCore.Utilities.Check.NotNull[T](T value, String parameterName) in /home/caleb/Projects/caleb/EntityFramework/src/Shared/Check.cs:line 27
   at Microsoft.EntityFrameworkCore.Storage.RelationalConnectionDependencies..ctor(IDbContextOptions contextOptions, IDiagnosticsLogger`1 transactionLogger, IDiagnosticsLogger`1 connectionLogger, INamedConnectionStringResolver connectionStringResolver) in /home/caleb/Projects/caleb/EntityFramework/src/EFCore.Relational/Storage/RelationalConnectionDependencies.cs:line 54
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.MySqlServiceCollectionExtensions.<>c.<AddEntityFrameworkMySql>b__0_1(IServiceProvider p) in /home/caleb/Projects/caleb/Pomelo.EntityFrameworkCore.MySql/src/EFCore.MySql/Extensions/MySqlServiceCollectionExtensions.cs:line 45
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService[T](IServiceProvider provider)
   at Microsoft.EntityFrameworkCore.Infrastructure.EntityFrameworkRelationalServicesBuilder.<>c.<TryAddCoreServices>b__3_1(IServiceProvider p) in /home/caleb/Projects/caleb/EntityFramework/src/EFCore.Relational/Infrastructure/EntityFrameworkRelationalServicesBuilder.cs:line 136
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
   at Microsoft.EntityFrameworkCore.Infrastructure.AccessorExtensions.GetService[TService](IInfrastructure`1 accessor) in /home/caleb/Projects/caleb/EntityFramework/src/EFCore/Infrastructure/AccessorExtensions.cs:line 51
   at Microsoft.EntityFrameworkCore.Infrastructure.DatabaseFacade.get_DatabaseCreator() in /home/caleb/Projects/caleb/EntityFramework/src/EFCore/Infrastructure/DatabaseFacade.cs:line 221
   at Microsoft.EntityFrameworkCore.Infrastructure.DatabaseFacade.EnsureCreated() in /home/caleb/Projects/caleb/EntityFramework/src/EFCore/Infrastructure/DatabaseFacade.cs:line 54
```

**Comment by ajcvickers** [comment]
@caleblloyd The issue is that MySqlMigrationsSqlGenerator is a singleton, but it depends on IRelationalConnection, which is scoped. Singleton services cannot depend on scoped services.  SQL generation in Migrations should work without a database connection--for example when generating scripts to later run on a production server.

**Comment by caleblloyd** [comment]
@ajcvickers that was it, removing `IRelationalConnection` from `MySqlMigrationsSqlGenerator` cleared this up

Wish there was an exception message or step-through debugging that was able to point to this.  It took many hours (days) to find this, only after you pointed it out :slightly_frowning_face: 

Thanks for the help on this, I'll keep chugging along upgrading the MySql provider.

**Comment by ajcvickers** [comment]
@caleblloyd In 2.0, the D.I. builder does have a way of verifying this, but it's not switched on by default and we haven't investigated enough that having it on by default when building a service provider in EF is the right thing to do. I'm going to do some more investigation into this.
